sdks/python/apache_beam/runners/direct/transform_evaluator.py

```diff
@@ -421,8 +440,12 @@ def process_element(self, element):
       main_output = list(self._outputs)[0]
       bundle = self._evaluation_context.create_bundle(main_output)
       for tv in event.timestamped_values:
-        bundle.output(
-            GlobalWindows.windowed_value(tv.value, timestamp=tv.timestamp))
+        # Unreify the value into the correct window.
+        try:
+          bundle.output(WindowedValue(**tv.value))
```

Comment:
This is fragile--we should not be (attempting to) interpret values as dicts of windowed value parameters and creating globally windowed values if anything goes wrong. Instead, store the WindowedValue object in tv.value. (Alternatively, one could let `events` be full WindowedValues rather than TimestampedValues, but that might be a bigger change.)

Reply:
(For that matter, it seems odd to be mutating ElementEvents in an operator called _WatermarkControllerEvaluator.)

Reply:
Unfortunately the Python SDK interprets an element that's a WindowedValue as an element that should have its windowing information changed. This is used directly by the ReverseTestStream to produce WindowedValue args that keep the windowing information. But if these elements were WindowedValues themselves, then the Python SDK would reinterpret and unnest them at every step.

Reply:
In that case, it'd be cleaner to make a new class WindowedValueHolder, or use a 1-tuple of WindowedValue, or something explicit like that rather than trying to use every element as kwargs to WindowedValue constructor and falling back otherwise.

Reply:
Alternatively, it might be cleaner to just pass the test stream event proto here rather than decoding it and then interpreting it. (Maybe this'd be a larger change.)

Reply:
Ack, for now I changed it to a WindowedValueHolder.